File: diameterOfBinaryTree.py
# Given the root of a binary tree, return the length of the diameter of the tree.
# The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.
# The length of a path between two nodes is represented by the number of edges between them.

class Solution(object):
    def getDepth(self, root):
        if not root:
            return 0
        
        leftDepth = self.getDepth(root.left)
        rightDepth = self.getDepth(root.right)
        
        return max(leftDepth, rightDepth) + 1

    def diameterOfBinaryTree(self, root):
        if not root:
            return 0       

        leftDiameter = self.diameterOfBinaryTree(root.left)
        rightDiameter = self.diameterOfBinaryTree(root.right)
        diameter = self.getDepth(root.left) + self.getDepth(root.right)
        diameter = max(diameter, max(leftDiameter, rightDiameter))

        return diameter

# Explanation
# Helper function to get depth (explained in previous code)
# Base case if root is empty return 0
# Recursively calculate the left and right subtrees
# Diameter is intialized to be the sum of the depth of the left and right subtrees
# Diameter is equal to the max num of the 3 numbers calculated
# Return diameter

File: diameterOfBinaryTree.c
// Given the root of a binary tree, return the length of the diameter of the tree.
// The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.
// The length of a path between two nodes is represented by the number of edges between them.

int maxDepth(struct TreeNode* root){
    if (root == NULL){
        return 0;
    }

    int leftDepth = maxDepth(root->left);
    int rightDepth = maxDepth(root->right);

    if (leftDepth > rightDepth){
        return leftDepth + 1;
    }
    else{
        return rightDepth + 1;
    }
}

int diameterOfBinaryTree(struct TreeNode* root){
    if (root == NULL){
        return 0;
    }
    int max = 0;
    int leftDiameter = diameterOfBinaryTree(root->left);
    int rightDiameter = diameterOfBinaryTree(root->right);
    int diameter = maxDepth(root->left) + maxDepth(root->right);
    if (leftDiameter > rightDiameter){
        max = leftDiameter;
    }
    else{
        max = rightDiameter;
    }

    if (max > diameter){
        return max;
    }
    else{
    return diameter;
    }
}

// Explanation
// MaxDepth function from previous code
// Base case to check if root is NULL
// Initialize a max variable for later
// Recurisvely calculate the left and right subtrees
// Initialize diameter as the sum of the depth of the left and right subtrees
// Compare left, right and diameter and return the largest
